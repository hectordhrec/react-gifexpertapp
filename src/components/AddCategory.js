import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        //setInputValue(e.target.value);
        setInputValue(valor => e.target.value);
        console.log('handleInputchange');
    }

    const handleSunmit = (e) => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories(categories => [inputValue,...categories]);
            setInputValue('');
        }
    }

    return (
        <form onSubmit={ handleSunmit }>
            <p>{inputValue}</p>
            <input type="text" className="form-control" value={inputValue} onChange={handleInputChange} />
        </form>
    );
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}

