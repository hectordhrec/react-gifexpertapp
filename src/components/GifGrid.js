import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {
    // const [images, setImages] = useState([]);

    /** Al colocar la categoría como dependencia, cada vez que la categoría cambie se va a ejecutar el useEffect */
    // useEffect(() => {
    //     getGifs(category).then(setImages);
    //     // getGifs().then(imgs => setImages(imgs)); // Otra forma
    // }, [category]);

    const { data: images, loading } = useFetchGifs(category);


    return (
        <>
            <h3 className="card animate__animated animate__fadeIn">{category}</h3>
            { loading && <p className="animate__animated animate__flash">Loading</p>}
            <div className="card-grid">
                {images.map((img) => (
                    <GifGridItem
                        key={img.id}
                        {...img}
                    />
                ))}
            </div>
        </>
    );
};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
