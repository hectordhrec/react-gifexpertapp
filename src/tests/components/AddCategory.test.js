import { AddCategory } from "../../components/AddCategory";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";

describe('Pruebas en el componente <AddCategory />', () => {

    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories = { setCategories } />);
    
    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories = { setCategories } />);
    });
    
    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('debe de cambiar la caja de texto', () => {
        const input = wrapper.find('input');

        const value = 'Hola mundo';
        input.simulate('change', { target: { value }});

        expect(wrapper.find('p').text().trim()).toBe(value);
    });

    test('No debe de postear con Submit', () => {
        wrapper.find('form').simulate('submit', {preventDefault(){}});

        expect(setCategories).not.toHaveBeenCalled();
    });

    test('Debe de llamar setCategories y limpiar la caja de texto', () => {
        const input = wrapper.find('input');
        const value = 'casa';
        // 1. simular el inputChange
        input.simulate('change', { target: { value }});

        // 2. simular el submit
        wrapper.find('form').simulate('submit', {preventDefault(){}});

        // 3. setCategories se debe haber llamado
        expect(setCategories).toHaveBeenCalled();
        // expect(setCategories).toHaveBeenCalledTimes(1); // para verificar que solo se haya ejecutado 1 vez en este caso
        // expect(setCategories).toHaveBeenCalledWith(expect.any(Function)); // para verificar que solo se haya ejecutado mandandole cualquier tipo de funci√≥n

        // 4. el valor del input debe ser ''
        expect(wrapper.find('input').prop('value')).toBe('');
    });
    
    
    
});
