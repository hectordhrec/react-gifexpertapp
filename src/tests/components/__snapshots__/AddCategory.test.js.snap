// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <AddCategory /> Debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    className="form-control"
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
